### PR TITLE
[#1197] Add "maxSlices" pie chart param to group slices into "Others"

### PR DESCRIPTION
**Pull request by vitorbaptista**
A pie chart doesn't work well if there're many categories, so this code adds a
"maxSlices" parameter to it where, if there're more slices than this number,
it'll group them in a single "Others" slice.

We do this by getting the totals from the "summary" attribute, and subtracting
the values of all slices that will still be shown.

This works OK but there's a caveat: it'll break if we're using anything other
tham a simple sum. For example, if the data contains averages we'll display
incorrect data.

In my opinion, this is OK, as a pie chart isn't well suited to display averages
anyway. We just need to take care not to use averages (or other summary methods)
when limiting the number of slices.

opentrials/opentrials#1197

### CODE REVIEW

**Comment by akariv** [comment]
Do you know why tests started failing?

File: src/components/pie/index.js [hunk 1]
```diff
@@ -90,14 +94,87 @@ export class PieChartComponent extends events.EventEmitter {
...
+    if (maxSlices <= 0) {
+      throw RangeError(`Can't create a pie chart with less than 1 slice (asked for ${maxSlices} slices)`);
+    }
+    if (!_.isArray(data.cells) || maxSlices === undefined || data.cells.length <= maxSlices) {
+      return data;
+    }
+
+    const individualCells = data.cells.slice(0, maxSlices - 1);
+    const othersCell = {
+      // Set all dimension's keys that end in "Value" to "Other"
```
Comment (akariv) [inline src/components/pie/index.js]: Where does this 'Value' constraint coming from?

File: src/components/pie/index.js [hunk 2]
```diff
@@ -90,14 +94,87 @@ export class PieChartComponent extends events.EventEmitter {
...
+        const theDimension = {};
+        Object.keys(dimension).forEach((dimensionKey) => {
+          const value = (dimensionKey.endsWith('Value')) ? this.i18n('others') : dimension[dimensionKey];
+          theDimension[dimensionKey] = value;
+        });
+        return theDimension;
+      }),
+
+      // Set measures as the total values minus the other cell's values
+      measures: Object.keys(data.summary).reduce((measures, measureKey) => {
```
Comment (akariv) [inline src/components/pie/index.js]: IMO simple for loops would've been more readable than using reduce here.

File: src/components/pie/index.js (hunk 1 above)
Comment (vitorbaptista) [inline src/components/pie/index.js]: This is my lack of understanding of OpenSpending's API. I assumed the `keyField` and `keyValue` came from the OpenSpending API, but just noticed that they're added by `babbage.ui` at https://github.com/openspending/babbage.ui/blob/0a7f9eb8006b804112febc6be5c95651a0137957/src/api/index.js#L306-L312. I'll change it.

File: src/components/pie/index.js (hunk 2 above)
Comment (vitorbaptista) [inline src/components/pie/index.js]: While reviewing the code, I figured out a way to simplify this code. Although it still uses a functional style, it's much more readable IMO. What do you think?

**Comment by vitorbaptista** [comment]
@akariv The tests were failing because the newer version of `jsdom` uses ES6 features not supported by the NodeJS 4.x we use. I downgraded `jsdom` and they're passing now. We should reserve some time later to upgrade NodeJS.

**Pull request merged**